cwrap/cfile.py

```diff
@@ -14,57 +14,158 @@
-        pythonapi which is ctypes wrapping of the CPython api.
+        def __del__(self):
+            pass
 
-          C-function:
-             void fprintf_hello(FILE * stream , const char * msg);
 
-          Python wrapper:
-             lib = clib.load( "lib.so" )
-             fprintf_hello = Prototype(lib, "void fprintf_hello( FILE , char* )")
+if six.PY3:
+    from .clib import load as cwrapload
 
-          Python use:
-             py_fileH = open("file.txt" , "w")
-             fprintf_hello( CFILE( py_fileH ) , "Message ...")
-             py_fileH.close()
+    class LibcPrototype(Prototype):
+        lib = cwrapload(None)
 
-        If the supplied argument is not of type py_file the function
-        will raise a TypeException.
+        def __init__(self, prototype, bind=False, allow_attribute_error=False):
+            super(LibcPrototype, self).__init__(
+                LibcPrototype.lib,
+                prototype,
+                bind=bind,
+                allow_attribute_error=allow_attribute_error)
 
-        Examples: ecl.ecl.ecl_kw.EclKW.fprintf_grdecl()
+    def copen(filename, mode='r'):
         """
-        c_ptr = self._as_file(py_file)
-        try:
-            super(CFILE, self).__init__(c_ptr)
-        except ValueError as e:
-            raise TypeError("Sorry - the supplied argument is not a valid Python file handle!")
+        This is a compatibility layer for functions taking FILE* pointers, and
+        should not be used unless absolutely needed.
+
+        In Python 2 this function is simply an alias for open. In Python 3,
+        however, it returns an instance of CWrapFile, a very lightweight
+        wrapper around a FILE* instance.
+        """
+        return CWrapFile(filename, mode)
+
+    class CWrapFile(BaseCClass):
+        """
+        This is a compatibility layer for functions taking FILE* pointers, and
+        should not be used unless absolutely needed.
+
+        CWrapFile is a very lightweight wrapper around FILE* instances. It is
+        meant be used inplace of python file objects that are to be passed to
+        foreign function calls under python 3.
+
+        Example:
+            with cwrap.open('filename', 'mode') as f:
+                foreign_function_call(f)
+        """
+
+        TYPE_NAME = "FILE"
+
+        _fopen = LibcPrototype("void* fopen (char*, char*)")
+        _fclose = LibcPrototype("int fclose (FILE)", bind=True)
+        _fflush = LibcPrototype("int fflush (FILE)", bind=True)
+
+        def __init__(self, fname, mode):
+            c_ptr = self._fopen(fname, mode)
+            self._mode = mode
+            self._fname = fname
+            self._closed = False
+
+            try:
+                super(CWrapFile, self).__init__(c_ptr)
+            except ValueError:
+                self._closed = True
+                raise IOError('Could not open file "{}" in mode {}'
+                              .format(fname, mode))
+
+        def close(self):
```

Comment:
so `_fclose` returns an `int` and we return `not _fclose()` unless we are closed, in which case we return `None`?

Reply:
The return value `0` from `fclose` means ok, by returning not of the result, we return a boolean that is true if we succeeded and false if not.

Reply:
And the None means it's already closed and thus had no effect?

Reply:
Oh yea, I guess that it should return True if already closed?